pkg/api/auth_middleware.go

```diff
@@ -148,11 +160,107 @@ func enhanceWithFriendlyName(user *model.User, friendlyName string) *model.User
 	return user
 }
 
+// userFromSAML returns a user from an existing SAML session.
+// If the user doesn't exist on the lakeFS side, it is created.
+// This function does not make any calls to an external provider.
+func userFromSAML(ctx context.Context, logger logging.Logger, authService auth.Service, authSession *sessions.Session, cookieAuthConfig *config.CookieAuthVerification) (*model.User, error) {
+	idTokenClaims, ok := authSession.Values[SAMLTokenClaimsSessionKey].(oidc_encoding.Claims)
+
+	if idTokenClaims == nil {
+		return nil, nil
+	}
+
+	if !ok {
+		logger.WithField("claims", authSession.Values[SAMLTokenClaimsSessionKey]).Error("failed decoding tokens")
+		return nil, ErrAuthenticatingRequest
+	}
+
+	logger.WithField("claims", idTokenClaims).Debug("success decoding token claims")
+
+	idKey := cookieAuthConfig.ExternalUserIDClaimName
+	externalID, ok := idTokenClaims[idKey].(string)
+	if !ok {
+		logger.WithField(idKey, idTokenClaims[idKey]).Error("failed type assertion for sub claim")
+		return nil, ErrAuthenticatingRequest
+	}
+
+	log := logger.WithField("external_id", externalID)
+
+	for claimName, expectedValue := range cookieAuthConfig.ValidateIDTokenClaims {
+		actualValue, ok := idTokenClaims[claimName]
+		if !ok || actualValue != expectedValue {
+			log.WithFields(logging.Fields{
+				"claim_name":     claimName,
+				"actual_value":   actualValue,
+				"expected_value": expectedValue,
+				"missing":        !ok,
+			}).Error("authentication failed on validating ID token claims")
+			return nil, ErrAuthenticatingRequest
+		}
+	}
+
+	// update user
+	// TODO(isan) consolidate userFromOIDC and userFromSAML below here internal db handling code
+	friendlyName := ""
+	if cookieAuthConfig.FriendlyNameClaimName != "" {
+		friendlyName, _ = idTokenClaims[cookieAuthConfig.FriendlyNameClaimName].(string)
+	}
+	log = log.WithField("friendly_name", friendlyName)
+
+	user, err := authService.GetUserByExternalID(ctx, externalID)
+	if err == nil {
+		log.Info("user found returning success ")
+		return enhanceWithFriendlyName(user, friendlyName), nil
+	}
+
+	if !errors.Is(err, auth.ErrNotFound) {
+		log.WithError(err).Error("failed while searching if user exists in database")
+		return nil, ErrAuthenticatingRequest
+	}
+	log.Info("user not found; creating them")
+
+	u := model.User{
+		CreatedAt:  time.Now().UTC(),
+		Source:     cookieAuthConfig.AuthSource,
+		Username:   externalID,
+		ExternalID: &externalID,
+	}
+
+	_, err = authService.CreateUser(ctx, &u)
+	if err != nil {
+		if !errors.Is(err, auth.ErrAlreadyExists) {
+			log.WithError(err).Error("failed to create external user in database")
+			return nil, ErrAuthenticatingRequest
+		}
+		// user already exists - get it:
+		user, err = authService.GetUserByExternalID(ctx, externalID)
```

Comment:
Don't we need the groups too in this flow?

Reply:
Nope, this is only initial registration, the first time the user ever get's registered to the system also I followed the OIDC blueprint :)